x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,329 @@
+            ttl
+          )} and token = ${this.token}`
+        );
+        return true;
+      } else if (response.result === 'noop') {
+        this.logger.debug(
+          `Lock "${this.lockId}" could not be acquired with token ${this.token} because it is already held`
+        );
+        return false;
+      } else {
+        throw new Error(`Unexpected response: ${response.result}`);
+      }
+    } catch (e) {
+      if (
+        e instanceof errors.ResponseError &&
+        e.body?.error?.type === 'version_conflict_engine_exception'
+      ) {
+        this.logger.debug(`Lock "${this.lockId}" already held (version conflict)`);
+        return false;
+      }
+
+      this.logger.error(`Failed to acquire lock "${this.lockId}": ${e.message}`);
+      this.logger.debug(e);
+      return false;
+    }
+  }
+
+  /**
+   * Releases the lock by deleting the document with the given lockId and token
+   */
+  public async release(): Promise<boolean> {
+    try {
+      const res = await pRetry(
+        () => {
+          return this.esClient.deleteByQuery({
+            index: LOCKS_CONCRETE_INDEX_NAME,
+            query: {
+              // bool: { filter: [{ term: { _id: this.lockId } }, { term: { token: this.token } }] },
+              bool: { filter: [{ term: { _id: this.lockId } }, { term: { token: this.token } }] },
+            },
+            refresh: true,
+          });
+        },
+        { retries: 2 }
+      );
+
+      if (res.deleted === 0) {
+        this.logger.debug(
+          `Lock "${this.lockId}" with token = ${this.token} could not be released. Not found or already released.`
+        );
+        return false;
+      }
+
+      this.logger.debug(`Lock "${this.lockId}" with token = ${this.token} was released.`);
+      return true;
+    } catch (error) {
+      this.logger.error(`Failed to release lock "${this.lockId}": ${error.message}`);
+      this.logger.debug(error);
+      return false;
+    }
+  }
+
+  /**
+   * Releases a lock if it is expired.
+   */
+  public async releaseIfExpired(): Promise<boolean> {
+    const res = await this.esClient.deleteByQuery({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { lte: 'now' } } }],
+        },
+      },
+      refresh: true,
+    });
+    if (res.deleted === 0) {
+      this.logger.debug(`Lock "${this.lockId}" not found`);
+      return false;
+    }
+
+    this.logger.debug(`Lock "${this.lockId}" expired and was released.`);
+    return true;
+  }
+
+  /**
+   * Retrieves the lock document for a given lockId.
+   * If the lock is expired, it will be released.
+   */
+  public async get(): Promise<LockDocument | undefined> {
+    const result = await this.esClient.search<LockDocument>({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { gt: 'now' } } }],
```

Comment:
clock skew

Reply:
As mentioned elsewhere, this is not uses anywhere atm. It's a helper for consumers to see whether a given lock is active. It should not be used in order to acquire a lock.